### PR TITLE
Add waitForMinimumCoordinators to ClusterSizeMonitor

### PR DESCRIPTION
**Pull request by tdcmeehan**
Test plan - Unit test coverage included in this PR

Extracted from #15071

Related to #10174

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 1]
```diff
@@ -131,22 +145,62 @@ public void stop()
...
     {
-        if (includeCoordinator) {
-            currentCount = allNodes.getActiveNodes().size();
-        }
-        else {
-            currentCount = Sets.difference(allNodes.getActiveNodes(), allNodes.getActiveCoordinators()).size();
-        }
+        currentCount = 0;
+        currentCoordinatorCount = 0;
+        allNodes.getActiveNodes().forEach(node -> {
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:187]: Any specific reason to not use the `getActiveNodes` and `getActiveCoordinator` methods here instead and again calculating it?

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 2]
```diff
@@ -131,22 +145,62 @@ public void stop()
...
+            if (node.isCoordinator()) {
+                currentCoordinatorCount += 1;
+                if (!includeCoordinator) {
+                    return;
+                }
+            }
+            currentCount += 1;
+        });
         if (currentCount >= executionMinCount) {
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:196]: Not sure how important it is, but if the cluster doesn't have enough coordinator and workers, we should always return not enough coordinator error. With current implementation if could be either one (not enough coordinator/not enough worker) of the error that we return and the behavior is not consistent. 
If we can guarantee that somehow, that would be great.

**Review by swapsmagic — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 3]
```diff
@@ -66,20 +73,27 @@ public ClusterSizeMonitor(InternalNodeManager nodeManager, NodeSchedulerConfig n
                 nodeManager,
                 requireNonNull(nodeSchedulerConfig, "nodeSchedulerConfig is null").isIncludeCoordinator(),
                 requireNonNull(queryManagerConfig, "queryManagerConfig is null").getRequiredWorkers(),
-                queryManagerConfig.getRequiredWorkersMaxWait());
+                queryManagerConfig.getRequiredWorkersMaxWait(),
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:76]: We do it above.  Can remove the check below.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java (hunk 1 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:187]: That's a good point, thanks.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 4]
```diff
@@ -131,22 +145,62 @@ public void stop()
...
+        // remove future if finished (e.g., canceled, timed out)
+        future.addListener(() -> {
+            timeoutTask.cancel(true);
+            removeFuture(future);
+        }, executor);
+
+        return future;
+    }
+
     private synchronized void removeFuture(SettableFuture<?> future)
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:178]: Good catch, thanks.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java (hunk 2 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:196]: We already have this behavior, because one will wait for one or the other via the `waitForMinimumWorkers` or `waitForMinimumCoordinators` methods.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 5]
```diff
@@ -66,20 +73,27 @@ public ClusterSizeMonitor(InternalNodeManager nodeManager, NodeSchedulerConfig n
...
     }
 
     public ClusterSizeMonitor(
             InternalNodeManager nodeManager,
             boolean includeCoordinator,
             int executionMinCount,
-            Duration executionMaxWait)
+            Duration executionMaxWait,
+            int minCoordinatorCount,
+            Duration maxCoordinatorWait)
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:87]: The inverted wording of the original annoyed me, but I can make them consistent.

**Comment by tdcmeehan** [comment]
@swapsmagic thanks for the review.  I addressed your comments.

**Review by swapsmagic — APPROVED** [review]
LGTM

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 6]
```diff
@@ -57,7 +59,13 @@
     private int currentCount;
```
Comment (mayankgarg1990) [inline presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:59]: should we rename this to `currentWorkerCount` to make it clearer?

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 7]
```diff
@@ -125,15 +140,50 @@ public void stop()
...
             timeoutTask.cancel(true);
-            removeFuture(future);
+            removeWorkerFuture(future);
         }, executor);
 
         return future;
     }
 
-    private synchronized void removeFuture(SettableFuture<?> future)
+    public synchronized ListenableFuture<?> waitForMinimumCoordinators()
```
Comment (mayankgarg1990) [inline presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:149]: instead of having 2 different functions - do you feel we should have a single function `waitForMinimumNodes()` (or some better name) which takes into account both workers and coordinators? Also, this PR does not use this function yet, will this be followed up in a separate PR? If yes, I am curious why?

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:149]: I would prefer to wait for them separately so we could have an accurate message indicating the resource problem (insufficient coordinators/insufficient workers).

You can see how it fits together in the `Extract from` reference in the PR description.  I'm putting these out as small PRs to make it easy to review thoroughly.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java (hunk 6 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:59]: Sure.

File: presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java [hunk 8]
```diff
@@ -125,28 +140,69 @@ public void stop()
...
+        if (currentWorkerCount >= workerMinCount) {
+            ImmutableList<SettableFuture<?>> listeners = ImmutableList.copyOf(workerSizeFutures);
+            workerSizeFutures.clear();
+            executor.submit(() -> listeners.forEach(listener -> listener.set(null)));
         }
-        if (currentCount >= executionMinCount) {
-            ImmutableList<SettableFuture<?>> listeners = ImmutableList.copyOf(futures);
-            futures.clear();
+        if (currentCoordinatorCount >= coordinatorMinCount) {
+            ImmutableList<SettableFuture<?>> listeners = ImmutableList.copyOf(coordinatorSizeFutures);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java:204]: `List<SettableFuture<?>>`

**Review by highker — APPROVED** [review]
lgtm
